File: sf15to18.py
def sf15to18 (id_):
	if not id_:
		raise ValueError('No id given.')
	if not isinstance(id_, str):
		raise TypeError('The given id isn\'t a string')
	if len(id_) == 18:
		return id_
	if len(id_) != 15:
		raise ValueError('The given id isn\'t 15 characters long.')

	# Generate three last digits of the id
	for i in range(3):
		f = 0

		# For every 5-digit block of the given id
		for j in range(5):
			# Assign the j-th chracter of the i-th 5-digit block to c
			c = id_[i * 5 + j]

			# Check if c is an uppercase letter
			if c.isupper():
				# Set a 1 at the character's position in the reversed segment
				f += 1 << j

		# Add the calculated character for the current block to the id
		id_ += 'ABCDEFGHIJKLMNOPQRSTUVWXYZ012345'[f]

	return id_

File: sf15to18.go
package sf15to18

import (
	"unicode"
	"bytes"
	"errors"
)

func Convert(id string) (string, error) {
	if len(id) == 18 {
		return id, nil
	}

	if len(id) != 15 {
		return "", errors.New("The given id isn't 15 characters long.")
	}

	const alphabet string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ012345"
	var longId bytes.Buffer

	// Write input id to output buffer as a starting point
	longId.WriteString(id)

	// Generate three last digits of the id
	for i := 0; i < 3; i++ {
		var f int = 0;

		// For every 5-digit block of the given id
		for j := 0; j < 5; j++ {
			// Assign the j-th chracter of the i-th 5-digit block to c
			var c rune = rune(id[i * 5 + j]);

			// Check if c is an uppercase letter
			if unicode.IsUpper(c) {
				// Set a 1 at the character's position in the reversed segment
				f += 1 << uint(j);
			}
		}

		// Add the calculated character for the current block to the id
		longId.WriteString(string(alphabet[f]));
	}

	return longId.String(), nil
}
